GLSP-1363 Refactor CLI application

tortmayr opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-glsp/glsp/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See [SECURITY.md](https://github.com/eclipse-glsp/glsp/blob/master/SECURITY.md),
to learn how to report vulnerabilities.
-->

#### What it does
Remove dependency to shelljs for CLI application, refactor release/releng commands and introduce bundling with esbuild

## Remove dependency to shelljs
Shelljs  has some issues in its core architecture which prevent from proper bundling.
We only use a very small subset of shelljs features. Therefore I opted for removing the dependency alltogether and introduce replacement functions in our utility library that directly use node builtins under the hood

## Refactor release/releng commands
- Remove the existing release commands and introduce a new set of releng commands that
  can also be used in CI and enable (partly) automated releases
- These commands have already been used for the 2.5.0 release (i.e. have been tested and work)
- Introduce helpers for working with (reading, updating etc.) workspace package jsons

## Bundle application with es build
- Introduce bundling with esbuild for the CLI package
- Package is now dependency free and ships the CLI app in one bundle (+ source maps)
- Update package scripts accordingly
- Introduce proper watch mode for CLI package
- Introduce vs code lauch config for debugin the CLI application. When starting the config the desired arguments/commands
 can be passed via input box

## Add workflows
- Add dispatch workflow to trigger a new release preparation PR for a repository

Fixes eclipse-glsp/glsp/issues/1363

<!-- Include relevant issues and describe how they are addressed. -->

#### How to test

<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Changelog

<!-- Please check, when if it applies to your change. -->

- [x] This PR should be mentioned in the changelog
- [ ] This PR introduces a breaking change (if yes, provide more details below for the changelog and the migration guide)

tortmayr committed:
GLSP-1363 Refactor CLI application

Remove dependency to shelljs for CLI application, refactor
release/releng commands and introduce bundling with esbuild

## Remove dependency to shelljs
Shelljs  has some issues in its core architecture which prevent it from
proper bundling.
We only use a very small subset of shelljs features.
Therefore I opted for removing the dependency alltogether and introduce
replacement functions in our utility library
that directly use node builtins under the hood

## Refactor release/releng commands
- Remove the existing release commands and introduce a new set of releng
commands that
  can also be used in CI and enable (partly) automated releases
- These commands have already been used for the 2.5.0 release (i.e. have
been tested and work)
- Introduce helpers for working with (reading, updating etc.) workspace
package jsons

## Bundle application with es build
- Introduce bundling with es build for the CLI package
- Package is now dependency free and ships the CLI app in one bundle (+
source maps)
- Update package scripts accordingly
- Introduce proper watch mode for CLI package
- Introduce vs code lauch config for debugin the CLI application. When
starting the config the desired arguments/commands
 can be passed via input box

## Add workflows
- Add dispatch workflow to trigger a new release preparation PR for a
repository

Fixes eclipse-glsp/glsp/issues/1363

tortmayr committed:
Update prepare-release.yml

martin-fleck-at commented on dev-packages/cli/src/commands/releng/prepare.ts:
```diff
@@ -0,0 +1,247 @@
+    .addArgument(new Argument('<versionType>', 'The version type').choices(VersionType.choices))
+    .argument('[customVersion]', 'Custom version number. Will be ignored if the release type is not "custom"')
+    .option('-v, --verbose', 'Enable verbose (debug) log output', false)
+    .option('-r, --repoDir <repoDir>', 'Path to the  component repository', validateGitDirectory, process.cwd())
```
```suggestion
    .option('-r, --repoDir <repoDir>', 'Path to the component repository', validateGitDirectory, process.cwd())
```

martin-fleck-at commented on dev-packages/cli/src/commands/releng/publish.ts:
```diff
@@ -0,0 +1,125 @@
+    .name('publish')
+    .description('Publish a new release for a GLSP component (npm, maven, github ...)')
+    .option('-v, --verbose', 'Enable verbose (debug) log output', false)
+    .option('-r, --repoDir <repoDir>', 'Path to the  component repository', validateGitDirectory, process.cwd())
```
```suggestion
    .option('-r, --repoDir <repoDir>', 'Path to the component repository', validateGitDirectory, process.cwd())
```

martin-fleck-at commented on dev-packages/cli/README.md:
```diff
@@ -105,25 +100,135 @@ Use this command to create an index file of all sources for a given directory an
+
+Options:
+  -v, --verbose            Enable verbose (debug) log output (default: false)
+  -r, --repoDir <repoDir>  Path to the  component repository (default:
```
```suggestion
  -r, --repoDir <repoDir>  Path to the component repository (default:
```

martin-fleck-at commented on dev-packages/cli/README.md:
```diff
@@ -105,25 +100,135 @@ Use this command to create an index file of all sources for a given directory an
+
+Options:
+  -v, --verbose            Enable verbose (debug) log output (default: false)
+  -r, --repoDir <repoDir>  Path to the  component repository (default:
```
```suggestion
  -r, --repoDir <repoDir>  Path to the component repository (default:
```

martin-fleck-at reviewed (approved):
Change looks good to me besides one minor formatting issue being copy-pasted ;-)

tortmayr committed:
Fix typos

Fix typos in command description
Ensure that non-glsp workspace dependencies also get properly updated
when using the version command
